README.md
# Print HTML Element

Prints the HTML of a provided element.

Originally forked from [jQuery.printElement](https://github.com/erikzaadi/jQueryPlugins/tree/master/jQuery.printElement) to remove jQuery dependency.

This fork removes some features while adding others (IE8 support is also removed for now). More features will be added as needed or requested.
PR's welcome.

View examples [here](https://rpdasilva.github.io/print-html-element/)

## Installation
Print HTML Element is available on both NPM and Bower.

`npm install print-html-element`

`bower install print-html-element`


## Usage


### Include
In a CommonJS module system via Browserify or Webpack
```js
    var PHE = require("print-html-element");
```

Good old-fashioned way
```html
    <script type="text/javascript" src="print-html-element.js"></script>
```
```js
    // Alias global variable printHtmlElement for purposes of example
    var PHE = printHtmlElement;
```

### Examples
View examples [here](https://rpdasilva.github.io/print-html-element/)

```js
    PHE.printElement( document.getElementById('toPrint') );
    PHE.printHtml("<h1>Let's print this h1</h1>");

    PHE.printHtml("<h1>Let's print this h1</h1>", {templateString: "<header>I'm part of the template header</header>{{printBody}}<footer>I'm part of the template footer</footer>"});
```

An HTML class `pe-body` is also added to the body of the print area which can be used as an additional style hook (on top of the regular print media query/stylesheet)

### Options and methods supported
```js
    var opts = {
        printMode: 'iframe',
        pageTitle: '',
        templateString: '',
        popupProperties: ''
    };

    PHE.printElement( elem, opts ); // Prints a DOM Element
    PHE.printHtml( str, opts ); // Prints an HTML string
```

- printMode determines which method is used to print. As a hidden `iframe` (default), or `popup` window
- pageTitle sets the printed page title (defaults to blank)
- templateString allows you to define a template that the html will be printed within.
    - Use `{{printBody}}` within the template to signify where the HTML should be injected
- popupProperties set the window features (such as `menubar`, `scrollbars`, etc. in `popup` mode


## Possible future features

- Integrate a template system (such as handlebars)
- Support multiple and/or custom template variables
- Support for multiple elements/HTML strings
- Promise/callback support


## License

`print-html-element` is [MIT licensed](LICENSE.txt)

CHANGELOG.md
<a name="0.5.1"></a>
# 0.4.2 (2016-06-28)

### Features
- **opts:** Add support for `opts.stylesheets` and `opts.styles` overrides

<a name="0.4.0"></a>
# 0.4.0 (2016-04-12)

### Features
- **bundle:** Converted to UMD module
- **build:** Build via `webpack`, `babel-loader` setup for planned migration to ES2015

### BREAKING CHANGES
- Transpiled and minified files now moved under `dist` folder. `<script>` tags may need to be updated
- `PrintElement` factory is no longer publie and is now called `PrintHtmlElement`
- When included outside of any module system (such as via `<script>` tag in browser), `PrintHtmlElement` is now constructed as global `printHtmlElement`

src/print-html-element.js
/*
* Print HTML Element
*
* Copyright (c) 2015 Philip Da Silva
*
* Forked from jQuery.printElement (https://github.com/erikzaadi/jQueryPlugins/tree/master/jQuery.printElement)
*
* Licensed under the MIT license:
*   http://www.opensource.org/licenses/mit-license.php
*/

function PrintHtmlElement() {
    function printElement(element, opts) {
        var elementHtml = element.outerHTML;

        _print(elementHtml, opts);
    }

    function printHtml(html, opts) {
        _print(html, opts);
    }

    function _print(html, opts) {
        opts = opts || {};
        opts = {
            printMode: opts.printMode || '',
            pageTitle: opts.pageTitle || '',
            templateString: opts.templateString || '',
            popupProperties: opts.popupProperties || '',
            stylesheets: opts.stylesheets || null,
            styles: opts.styles || null
        };

        // Get markup to be printed
        var markup = _getMarkup(html, opts),
            printWindow,
            printIframe,
            printDocument,
            printElementID;

        if (opts.printMode.toLowerCase() === 'popup') {
            printWindow = window.open('about:blank', 'printElementWindow', opts.popupProperties);
            printDocument = printWindow.document;
        } else {
            //The random ID is to overcome a safari bug
            // http://www.cjboco.com.sharedcopy.com/post.cfm/442dc92cd1c0ca10a5c35210b8166882.html
            printElementID = 'printElement_' + (Math.round(Math.random() * 99999)).toString();

            printIframe = document.createElement('iframe');
            printIframe.setAttribute('id', printElementID);
            printIframe.setAttribute('src', 'about:blank');
            printIframe.setAttribute('frameBorder', '0');
            printIframe.setAttribute('scrolling', 'no');
            printIframe.setAttribute('style', 'position:fixed;bottom:100%;right:100%;');

            document.body.appendChild(printIframe);

            printDocument = (printIframe.contentWindow || printIframe.contentDocument);
            if (printDocument.document) {
                printDocument = printDocument.document;
            }

            printIframe = document.frames ? document.frames[printElementID] : document.getElementById(printElementID);
            printWindow = printIframe.contentWindow || printIframe;
        }

        focus();
        printDocument.open();

        // SetTimeout fixes Issue #9 (iframe printMode does not work in firefox)
        setTimeout(function() {
          printDocument.write(markup);
          printDocument.close();
        });

        _callPrint(printWindow, printIframe);
    }

    function _callPrint(printWindow, iframe) {
        if (printWindow && printWindow.printPage) {
            printWindow.printPage();

            if (iframe) {
                // Remove iframe after printing
                document.body.removeChild(iframe);
            }
        } else {
            setTimeout(function() {
                _callPrint(printWindow, iframe);
            }, 50);
        }
    }

    function _getBaseHref() {
        var port = (window.location.port) ? ':' + window.location.port : '';
        return window.location.protocol + '//' + window.location.hostname + port + window.location.pathname;
    }

    function _getMarkup(elementHtml, opts) {
        var template = opts.templateString,
            templateRegex = new RegExp(/{{\s*printBody\s*}}/gi),
            stylesheets,
            styles,
            html = [];

        if(template && templateRegex.test(template)) {
            elementHtml = template.replace(templateRegex, elementHtml);
        }

        html.push('<html><head><title>' + (opts.pageTitle || '') + '</title>');

        // If stylesheet URL's or list of stylesheet URL's are specified, override page stylesheets
        if (opts.stylesheets) {
            stylesheets = Array.isArray(opts.stylesheets) ? opts.stylesheets : [opts.stylesheets];
        } else {
            stylesheets = Array.prototype.slice
                .call(document.getElementsByTagName('link'))
                .map(function(link) {
                    return link.href;
                });
        }

        stylesheets.forEach(function(href) {
            html.push('<link rel="stylesheet" href="' + href + '">');
        });

        // If inline styles or list of inline styles are specified, override inline styles
        if (opts.styles) {
            styles = Array.isArray(opts.styles) ? opts.styles : [opts.styles];
        } else {
            styles = Array.prototype.slice
                .call(document.getElementsByTagName('style'))
                .map(function(style) {
                    return style.innerHTML;
                });
        }
 
        styles.forEach(function (style) {
            html.push('<style type="text/css">' + style + '</style>');
        });

        // Ensure that relative links work
        html.push('<base href="' + _getBaseHref() + '" />');
        html.push('</head><body class="pe-body">');
        html.push(elementHtml);
        html.push('<script type="text/javascript">function printPage(){focus();print();' + ((opts.printMode.toLowerCase() == 'popup') ? 'close();' : '') + '}</script>');
        html.push('</body></html>');

        return html.join('');
    }

    return {
        printElement: printElement,
        printHtml: printHtml
    };
};

module.exports = PrintHtmlElement();


print-element.js
/*
* Print HTML Element 0.2.1
*
* Copyright (c) 2015 Philip Da Silva
*
* Forked from jQuery.printElement (https://github.com/erikzaadi/jQueryPlugins/tree/master/jQuery.printElement)
*
* Licensed under the MIT license:
*   http://www.opensource.org/licenses/mit-license.php
*/

var PrintElement = function() {
    function printElement(element, opts) {
        _print(element, opts, 'elem');
    }

    function printHtml(str, opts) {
        _print(str, opts, 'html');
    }

    function _print(element, opts, type) {
        opts = opts || {};
        opts = {
            printMode: opts.printMode || '',
            pageTitle: opts.pageTitle || ''
        };

        // Get markup to be printed
        var html = _getMarkup(element, opts, type),
            printWindow,
            printIframe,
            printDocument,
            printElementID;

        if (opts.printMode.toLowerCase() == 'popup')
        {
            printWindow = window.open('about:blank', 'printElementWindow');
            printDocument = printWindow.document;
        }
        else
        {
            printElementID = 'printElement_' + (Math.round(Math.random() * 99999)).toString();

            printIframe = document.createElement('iframe');
            printIframe.setAttribute('id', printElementID); //The random ID is to overcome a safari bug http://www.cjboco.com.sharedcopy.com/post.cfm/442dc92cd1c0ca10a5c35210b8166882.html
            printIframe.setAttribute('src', 'about:blank');
            printIframe.setAttribute('frameBorder', '0');
            printIframe.setAttribute('scrolling', 'no');
            printIframe.setAttribute('style', 'position:fixed;bottom:100%;right:100%;');

            document.body.appendChild(printIframe);

            printDocument = (printIframe.contentWindow || printIframe.contentDocument);
            if (printDocument.document)
            {
                printDocument = printDocument.document;
            }

            printIframe = document.frames ? document.frames[printElementID] : document.getElementById(printElementID);
            printWindow = printIframe.contentWindow || printIframe;
        }

        focus();
        printDocument.open();
        printDocument.write(html);
        printDocument.close();

        _callPrint(printWindow, printIframe);
    }

    function _callPrint(element, iframe) {
        if (element && element.printPage)
        {
            element.printPage();
            document.body.removeChild(iframe); // Remove iframe afrer printing
        }
        else
        {
            setTimeout(function() {
                _callPrint(element, iframe);
            }, 50);
        }
    }

    function _getBaseHref() {
        var port = (window.location.port) ? ':' + window.location.port : '';
        return window.location.protocol + '//' + window.location.hostname + port + window.location.pathname;
    }

    function _getMarkup(element, opts, type) {
        var elementHtml = element.innerHTML, // Does not support form input elements
            stylesheets,
            html = [];

        switch(type)
        {
            case 'elem':
                elementHtml = element.innerHTML;
                break;

            case 'html':
                elementHtml = element;
                break;

            default:
                elementHtml = element.innerHTML;
        }

        html.push('<html><head><title>' + (opts.pageTitle || '') + '</title>');

        stylesheets = Array.prototype.slice.call( document.getElementsByTagName('link') );
        stylesheets.forEach(function(link){
            html.push('<link rel="stylesheet" href="' + link.href + '">');
        });

        // Ensure that relative links work
        html.push('<base href="' + _getBaseHref() + '" />');
        html.push('</head><body class="pe-body">');
        html.push('<div class="' + element.className + '">' + elementHtml + '</div>');
        html.push('<script type="text/javascript">function printPage(){focus();print();' + ((opts.printMode.toLowerCase() == 'popup') ? 'close();' : '') + '}</script>');
        html.push('</body></html>');

        return html.join('');
    }

    return {
        printElement: printElement,
        printHtml: printHtml
    };
};

if (typeof module === 'object' && module.exports === exports)
{
    module.exports = PrintElement();
}